gcc/rust/resolve/rust-ast-resolve-type.cc

```diff
@@ -492,20 +481,79 @@ ResolveTypeToCanonicalPath::ResolveTypeToCanonicalPath ()
   : ResolverBase (), result (CanonicalPath::create_empty ())
 {}
 
+bool
+ResolveGenericArgs::is_const_value_name (const CanonicalPath &path)
+{
+  NodeId resolved;
+  auto found = resolver->get_name_scope ().lookup (path, &resolved);
+
+  return found;
+}
+
+bool
+ResolveGenericArgs::is_type_name (const CanonicalPath &path)
+{
+  NodeId resolved;
+  auto found = resolver->get_type_scope ().lookup (path, &resolved);
+
+  return found;
+}
+
 void
-ResolveGenericArgs::go (AST::GenericArgs &args)
+ResolveGenericArgs::disambiguate (AST::GenericArg &arg)
 {
-  for (auto &arg : args.get_generic_args ())
+  auto path = canonical_prefix.append (
+    CanonicalPath::new_seg (UNKNOWN_NODEID, arg.get_path ()));
+
+  auto is_type = is_type_name (path);
+  auto is_value = is_const_value_name (path);
```

Comment:
Is this the cast that we have AST::GenericArg as GenericArg::EITHER? I am just thinking it could be a PathInExpression or TypePath which will need us to call the appropriate ResolvePath or ResolveTypePath. 

This will work for all primitive types and values but not for paths. Lets go with this for now because I think the overall logic is right but we need to look at the GenericArg structure but we can do that in a separate PR